### PR TITLE
Receive Support + Refactoring + Bugfix

### LINKED ISSUE #9: [Feature] Add ability to send animated gifs instead of snapshots
I did some experimenting tonight and established a good baseline for animated gifs as an attachment option.  I remember the telegram plugin doing something similar.  Sure enough, there is a very simple conversion feature built into ffmpeg (which is bundled with all octopi installs and most custom ones too):

You can produce some high quality animated gifs using encoding like shown below.
```
ffmpeg -t 10 -y -threads 1 -i http://octopi-s1pro/webcam/?action=stream -filter_complex "fps=23,scale=-1:720" output.gif
```
Unfortunately to get it within Github's 10mb limit, I had to scale it back a bit:
output
```
ffmpeg -t 4 -y -threads 1 -i http://octopi-s1pro/webcam/?action=stream -filter_complex "fps=10,scale=-1:720" output.gif
```

### PR DESCRIPTION
**Pull request by synman**
-Group by Machine preference
-Supports json-rpc and websocket for receiving messages
-Numerous commands supported (see screenshot)
-Comma delimited list of progress values to send updates at
-Bugfix for PRINT_FAILED event missing reason tag
-Refactored / Reorganized code to match flow
-fixes #9 by adding ability to send animated gifs
-will only respond to known senders

Screenshot 2023-01-23 at 10 10 03 PM

### CODE REVIEW

**Comment by synman** [comment]
I will be amending this PR.   I will be removing the "try again" mitigation I put in place for single account / multiple machines use cases.   It is creating some weird results when you have a single account across multiple signal-cli instances / containers in use.

**Comment by synman** [comment]
I have one more round of testing to do tonight and assuming all goes well, I'll move this out of draft.   I'm running out of things to add.

File: octoprint_signalclirestapi/__init__.py [hunk 1]
```diff
@@ -241,10 +454,97 @@ def _send_message(self, message):
...
+                self._send_message(message)
+
+    def on_demand_status_report(self):
+        if self.enabled:
+            # populate tags managed elsewhere
+            self.update_supported_tags()
+            tags = self._supported_tags.copy()
+
+            if self._printer.is_printing() or self._printer.is_paused() or self._printer.is_pausing():
+                tags["progress"] = "{}%}".format(self._supported_tags["progress"])
```
Comment (synman) [inline octoprint_signalclirestapi/__init__.py:534]: typo

**Comment by synman** [comment]
@bbernhard - This PR is good to go... ready for input / feedback / review

**Comment by bbernhard** [comment]
Many thanks for the PR!

I just went over the changes and they look really good! I'll do a bit of testing in the next days and then merge the PR!

Thanks a lot!

**Comment by synman** [comment]
> Many thanks for the PR!
> 
> I just went over the changes and they look really good! I'll do a bit of testing in the next days and then merge the PR!
> 
> Thanks a lot!

least I can do.  thank you for providing such an awesome foundation to build upon -- both the plugin and the restful api / docker container.

I have an additional PR lined up that adds animated gif support.  I can merge that here or hold off as I've got it in a feature branch based on this one.

I also want to make group creation a bit cleaner.   There's a bug there I labeled as a `TODO`

**Comment by bbernhard** [comment]
> 

Thanks for the kind words! 

For me both is fine, so whatever works best for you. If you want to extend this PR, please feel free to do so :)

**Comment by synman** [comment]
I'll push a commit to this PR tonight with the animated gif feature and a reworked instance of the receive thread.

I'm not happy with how my implementation currently responds (it doesn't) to things like the sender phone number changing without requiring a restart.

**Comment by synman** [comment]
shoot... I forgot to reference "fixed #9 " in my commit

**Comment by synman** [comment]
I have one more commit to push tonight and then I'm REALLY REALLY done.   

I just realized we need some basic security on the receive side of things so I'm locking down processing of received messages to only senders who are on our recipients list.

**Comment by synman** [comment]
ok @bbernhard  I'm really done now :)

**Comment by bbernhard** [comment]
Thanks a lot! Awesome work!

**Comment by synman** [comment]
I threw together a teaser ... wasn't sure how you'd feel it about it, but it's here if you want it (I haven't promoted it all that much up to this point):  https://youtu.be/UjMrgx2to1s

**Comment by bbernhard** [comment]
That's really nice, thanks for sharing! I'll add a link to your video in the github release notes when I create the new release tag.

If you want to, we can also add that to the `README.md` file. Feel free to create a PR for that :)

**Comment by synman** [comment]
we should probably also add some notes about the performance gains achieved with `json-rpc `enabled.

**Pull request merged**